bitsail-connectors/bitsail-connectors-legacy/bitsail-connector-kafka/src/test/java/com/bytedance/bitsail/connector/legacy/kafka/source/KafkaSourceITCase.java

```diff
@@ -59,34 +68,100 @@ private static String constructARecord(int index) {
     kafkaCluster.startService();
     kafkaCluster.createTopic(topicName);
-    startSendDataToKafka();
   }
 
-  private void startSendDataToKafka() {
+  private void startSendJsonDataToKafka() {
     KafkaProducer<String, String> producer = kafkaCluster.getProducer(topicName);
     ScheduledThreadPoolExecutor produceService = new ScheduledThreadPoolExecutor(1);
     AtomicInteger sendCount = new AtomicInteger(0);
     produceService.scheduleAtFixedRate(() -> {
       try {
-        for (int i = 0; i < 5000; ++i) {
-          String record = constructARecord(sendCount.getAndIncrement());
-          producer.send(new ProducerRecord(topicName, record));
-        }
+	for (int i = 0; i < 5000; ++i) {
+	  String record = constructARecord(sendCount.getAndIncrement());
+	  producer.send(new ProducerRecord(topicName, record));
+	}
       } catch (Exception e) {
-        LOG.error("failed to send a record");
+	LOG.error("failed to send a record");
       } finally {
-        LOG.info(">>> kafka produce count: {}", sendCount.get());
+	LOG.info(">>> kafka produce count: {}", sendCount.get());
+      }
+    }, 0, 1, TimeUnit.SECONDS);
+  }
+
+  private void startSendPbDataToKafka(BitSailConfiguration configuration) throws Exception {
+    PbBytesParser parser = new PbBytesParser(configuration);
+    List<Descriptors.FieldDescriptor> fields = parser.getDescriptor().getFields();
+    Descriptors.Descriptor type = parser.getDescriptor();
+
+    KafkaProducer<String, String> producer = kafkaCluster.getProducer(topicName);
+    ScheduledThreadPoolExecutor produceService = new ScheduledThreadPoolExecutor(1);
+    AtomicInteger sendCount = new AtomicInteger(0);
+    produceService.scheduleAtFixedRate(() -> {
+      try {
+	for (int i = 0; i < 5000; ++i) {
+	  DynamicMessage.Builder builder = DynamicMessage.newBuilder(type);
+	  for (Descriptors.FieldDescriptor field : fields) {
+	    switch (field.getJavaType()) {
+	      case INT:
+		builder.setField(field, i);
+		break;
+	      case LONG:
+		builder.setField(field, (long) i);
+		break;
+	      case FLOAT:
+		builder.setField(field, (float) i);
+		break;
+	      case DOUBLE:
+		builder.setField(field, (double) i);
+		break;
+	      case BOOLEAN:
+		builder.setField(field, i % 2 == 0);
+		break;
+	      case BYTE_STRING:
+		builder.setField(field, ("byte_" + i).getBytes());
+		break;
+	      case MESSAGE:
+	      case STRING:
+	      case ENUM:
+	      default:
+		builder.setField(field, "text_" + i);
+		break;
+	    }
+	  }
+	  DynamicMessage message = builder.build();
+	  producer.send(new ProducerRecord(topicName, new String(message.toByteArray())));
+	  sendCount.getAndIncrement();
+	}
+      } catch (Exception e) {
+	LOG.error("failed to send a record");
+      } finally {
+	LOG.info(">>> kafka produce count: {}", sendCount.get());
       }
     }, 0, 1, TimeUnit.SECONDS);
   }
 
   @Test
-  public void testKafkaSource() throws Exception {
+  public void testKafkaSourceJsonFormat() throws Exception {
```

Comment:
I ran this ITCase and one test was failed. looks like the kafka producer didn't stop when we switch to the second test.

Reply:
Added clean up logic on the producer service to stop creating records. All test cases pass now

![image](https://user-images.githubusercontent.com/20527912/203223409-134ab769-f979-4e00-bdd3-499e6069dbab.png)